bitvm/src/signatures/public.rs

```diff
@@ -0,0 +1,838 @@
+use bitcoin::hex::DisplayHex;
+use bitcoin_script::Script;
+
+use super::utils::u32_to_le_bytes_minimal;
+use crate::signatures::winternitz;
+use crate::signatures::winternitz::{
+    BruteforceVerifier, Converter, ListpickVerifier, Parameters, ToBytesConverter, VoidConverter,
+    Winternitz,
+};
+
+/// Secret key for Winternitz signatures.
+///
+/// The same key type is used for all message lengths.
+pub type WinternitzSecret = winternitz::SecretKey;
+
+/// Public key for some Winternitz signature verification algorithm.
+///
+/// The key has to be converted into the right length before it can be used
+/// in any algorithm. The conversion might fail.
+pub type GenericWinternitzPublicKey = winternitz::PublicKey;
+
+/// Bundles a message with a secret key for signing.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct WinternitzSigningInputs<'a, 'b, WOTS: Wots + ?Sized> {
+    pub message: &'a WOTS::Message,
+    pub signing_key: &'b WinternitzSecret,
+}
+
+/// Number of bits per digit.
+///
+/// We hardcode the base to be 16. Therefore, there are 4 bits.
+pub const LOG2_BASE: u32 = 4;
+
+/// High-level functionality for working with Winternitz signatures.
+///
+/// Signatures contain the signature of each digit as well as the digit itself.
+///
+/// ## See
+///
+/// [`CompactWots`]
+pub trait Wots {
+    type Converter: Converter;
+    type PublicKey: AsRef<[[u8; 20]]> + TryFrom<Vec<[u8; 20]>, Error: std::fmt::Debug>;
+    type Message: AsRef<[u8]> + TryFrom<Vec<u8>, Error: std::fmt::Debug>;
+    type Signature: AsRef<[[u8; 21]]> + TryFrom<Vec<[u8; 21]>, Error: std::fmt::Debug>;
+
+    const ALGORITHM: Winternitz<ListpickVerifier, Self::Converter> = Winternitz::new();
+    const MSG_BYTE_LEN: u32;
+    const PARAMETERS: Parameters = Parameters::new_by_bit_length(Self::MSG_BYTE_LEN * 8, LOG2_BASE);
+    const TOTAL_DIGIT_LEN: u32 = Self::PARAMETERS.total_digit_len();
+
+    /// Generates a random secret key.
+    fn generate_secret_key() -> WinternitzSecret {
+        let mut buffer = [0u8; 20];
+        let mut rng = rand::rngs::OsRng;
+        rand::RngCore::fill_bytes(&mut rng, &mut buffer);
+        Vec::from(buffer)
+    }
+
+    /// Creates a secret key from the given `secret` string.
+    ///
+    /// ## Warning
+    ///
+    /// For backwards compatibility, the original conversion function is used.
+    /// The `secret` string is converted into ASCII bytes,
+    /// which are in turn converted into lower hex ASCII bytes.
+    ///
+    /// Usage of this method is generally discouraged.
+    /// It is better to create a [`WinternitzSecret`] via [`Vec<u8>`].
+    fn secret_from_str(secret: &str) -> WinternitzSecret {
+        secret.as_bytes().to_lower_hex_string().into_bytes()
+    }
```

Comment:
Add a `deprecated` in this unutterable cursed ugly mess:

```rs
#[deprecated(since="X.Y.Z", note="please use `new_method` instead")]
```

Reply:
The method is actually not used in the code. I added it to make it easier to port the bridge code to the new API, which turned out to be too much work.

I can remove the method from `public.rs` and add a warning + deprecation to `signing_winternitz.rs`.